qp_shotgun/humann2/humann2.py

```diff
@@ -11,23 +11,225 @@
 
 from future.utils import viewitems
 from functools import partial
+from subprocess import Popen
+from contextlib import contextmanager
+from tempfile import mkdtemp
+from shutil import rmtree
 
 from qiita_client import ArtifactInfo
 from qiita_client.util import system_call, get_sample_names_by_run_prefix
 
 
-def generate_humann2_analysis_commands(forward_seqs, reverse_seqs, map_file,
-                                       out_dir, parameters):
-    """Generates the HUMAnN2 commands
+@contextmanager
+def make_temp_directory():
+    temp_dir = mkdtemp()
+    try:
+        yield temp_dir
+    finally:
+        rmtree(temp_dir)
+
+
+def make_read_sets_per_sample(files, map_file):
+    """Recovers read set information from kneaddata output
 
     Parameters
     ----------
-    forward_seqs : list of str
-        The list of forward seqs filepaths
-    reverse_seqs : list of str
-        The list of reverse seqs filepaths
+    files : list of str
+        The list of sequence filepaths from the kneaddata artifact
     map_file : str
         The path to the mapping file
+
+    Returns
+    -------
+    read_sets: list of tup
+        list of 7-tuples with run prefix, sample name, fwd paired read fp,
+        rev paired read fp, fwd unpaired read fp, rev unpaired read fp, and
+        single fwd read fp.
+
+    Raises
+    ------
+    ValueError
+        If there are files matching the kneaddata paired file naming convention
+        (_paired_1.fastq.gz, _unmatched_1.fastq.gz) but which don't have all 4
+        outputs.
+    ValueError
+        If there are files matching the kneaddata paired file naming convention
+        (_paired_1.fastq.gz, _unmatched_1.fastq.gz) in addition to fastq.gz
+        files that do not match naming convention (the latter are interpreted
+        as single read files).
+    ValueError
+        If there are no *.fastq.gz files in the artifact
+
+    Notes
+    -----
+    """
+
+    # sort through the filenames and bin into sequence type lists
+    fwd_paired = []
+    fwd_unpaired = []
+    rev_paired = []
+    rev_unpaired = []
+    single = []
+
+    for fp in files:
+        if fp.endswith('_paired_1.fastq.gz'):
+            fwd_paired.append(fp)
+        elif fp.endswith('_paired_2.fastq.gz'):
+            rev_paired.append(fp)
+        elif fp.endswith('_unmatched_1.fastq.gz'):
+            fwd_unpaired.append(fp)
+        elif fp.endswith('_unmatched_2.fastq.gz'):
+            rev_unpaired.append(fp)
+        elif fp.endswith('.fastq.gz'):
+            single.append(fp)
+
+    # check that seq lists are same len
+    if not (len(fwd_paired) == len(fwd_unpaired) ==
```

Comment:
Just noticed this but you are reusing the len(array) many times and it could be better to calculate once and just reuse the values, for example:
lfp = len(fwd_paired)